plugins/repository-hdfs/src/main/java/org/elasticsearch/repositories/hdfs/HdfsBlobContainer.java

```diff
@@ -90,20 +91,20 @@ public InputStream readBlob(String blobName) throws IOException {
             throw new NoSuchFileException("Blob [" + blobName + "] does not exist");
         }
         // FSDataInputStream does buffering internally
-        return store.execute((Operation<InputStream>) fileContext -> {
+        return store.execute(fileContext -> {
             FSDataInputStream is = fileContext.open(new Path(path, blobName), bufferSize);
-            return new InputStream() {
-                @Override
-                public int read() throws IOException {
-                    try {
-                        SpecialPermission.check();
-                        // FSDataInputStream can open connection on read()
-                        return AccessController.doPrivileged((PrivilegedExceptionAction<Integer>) is::read);
-                    } catch (PrivilegedActionException e) {
-                        throw (IOException) e.getCause();
-                    }
-                }
-            };
+            InputStream stream = is.markSupported() ? is : new BufferedInputStream(is);
```

Comment:
if we run into an exception here we have to close the stream. 

we usually do this:

```Java
boolean success = false;
try {
 // do something with the stream
  success = true;
  return stream;
} finally {
  if (success == false) {
    IOUtils.closeWhileHandlingException(stream);
  }
}